storage/aptosdb/src/pruner/state_store/mod.rs

```diff
@@ -130,38 +126,27 @@ impl StateStorePruner {
         const MIN_INTERVAL: Duration = Duration::from_secs(10);
         const MIN_VERSIONS: u64 = 60000;
 
+        let min_readable_version = self.min_readable_version.load(Ordering::Relaxed);
+        let index_purged_until_version = self.index_purged_until_version.load(Ordering::Relaxed);
+
         // A deletion is issued at most once in one minute and when the pruner has progressed by at
         // least 60000 versions (assuming the pruner deletes as slow as 1000 versions per second,
         // this imposes at most one minute of work in vain after restarting.)
         let now = Instant::now();
-        if self.min_readable_version.load(Ordering::Relaxed) < self.index_min_nonpurged_version() {
-            warn!("State pruner inconsistent, min_readable_version is {} and  index_min_non-purged_version is {}",
-                self.min_readable_version.load(Ordering::Relaxed), self.index_min_nonpurged_version());
-            return Ok(());
-        }
-
         if now - *self.index_purged_at.lock() > MIN_INTERVAL
-            && self.min_readable_version.load(Ordering::Relaxed)
-                - self.index_min_nonpurged_version()
-                + 1
-                > MIN_VERSIONS
+            && min_readable_version - index_purged_until_version > MIN_VERSIONS
         {
-            let new_min_non_purged_version = self.min_readable_version.load(Ordering::Relaxed) + 1;
```

Comment:
This is correct, because the stale node index items are already dealt with at min_readable_version, so the next version where we need to keep the index items unpurged should be that plus 1.

However the if condition did the subtraction before +1, which can be a underflow.

It used not to trip before the state checkpoint thing because there most likely is some pruning happens in 10 seconds so min_readable_version becomes  1+ larger. But with the change (and pruning_batch_size==500), there can be only a few repeated empty state pruning batches that doesn't bump the number in the entire 10 seconds.. especially when sometimes the ledger pruner is slow (with the txn accumulator pruner doing occasional large commits).

Reply:
For my understanding, why don't we maintain the invariant that `min_readable_version` is always >= `index_min_nonpurged_version`? IIUC `min_readable_version` > version pruned = `index_min_nonpurged_version - 1`. My question is why would the underflow happen in the first place.